In [1]:
import pandas as pd
import numpy as np

In [2]:
df_2018 = pd.read_excel('MasterDataTracer.xlsx', sheet_name='2018') 
df_2019 = pd.read_excel('MasterDataTracer.xlsx', sheet_name='2019') 
df_2020 = pd.read_excel('MasterDataTracer.xlsx', sheet_name='2020') 
df_2021 = pd.read_excel('MasterDataTracer.xlsx', sheet_name='2021') 
df_2022 = pd.read_excel('MasterDataTracer.xlsx', sheet_name='2022') 

# df_reindex = pd.concat([df1, df2, df3, df4, df5], ignore_index=True)
# df_reindex.head()

# **Tambah kolom tahun survey**

In [3]:
df_2018['Tahun Survey'] = 2018
df_2019['Tahun Survey'] = 2019
df_2020['Tahun Survey'] = 2020
df_2021['Tahun Survey'] = 2021
df_2022['Tahun Survey'] = 2022

In [ ]:
# df_dict = pd.read_excel('MasterDataTracer.xlsx',sheet_name=None)

# {
#     '2018' : pd.DataFrame,
#     '2019' : pd.DataFrame,
#     '2020' : pd.DataFrame,
#     '2021' : pd.DataFrame,
#     '2022' : pd.DataFrame
# }

# df_2018 = df_dict.get('2018')
# df_2019 = df_dict.get('2019')
# df_2020 = df_dict.get('2020')
# df_2021 = df_dict.get('2021')
# df_2022 = df_dict.get('2022')

# **Check Jumlah Data**

In [4]:
def hitung_jumlah_kolom(dataframe):
    jumlah_kolom = len(dataframe.columns)
    return jumlah_kolom

In [5]:
# Menggunakan fungsi untuk menghitung jumlah kolom dalam df_2018
jumlah_kolom_df_2018 = hitung_jumlah_kolom(df_2018)
print(f"Jumlah kolom dalam df_2018: {jumlah_kolom_df_2018}")

# Menggunakan fungsi untuk menghitung jumlah kolom dalam df_2019
jumlah_kolom_df_2019 = hitung_jumlah_kolom(df_2019)
print(f"Jumlah kolom dalam df_2019: {jumlah_kolom_df_2019}")

# Menggunakan fungsi untuk menghitung jumlah kolom dalam df_2020
jumlah_kolom_df_2020 = hitung_jumlah_kolom(df_2020)
print(f"Jumlah kolom dalam df_2020: {jumlah_kolom_df_2020}")

# Menggunakan fungsi untuk menghitung jumlah kolom dalam df_2021
jumlah_kolom_df_2021 = hitung_jumlah_kolom(df_2021)
print(f"Jumlah kolom dalam df_2021: {jumlah_kolom_df_2021}")

# Menggunakan fungsi untuk menghitung jumlah kolom dalam df_2022
jumlah_kolom_df_2022 = hitung_jumlah_kolom(df_2022)
print(f"Jumlah kolom dalam df_2022: {jumlah_kolom_df_2022}")

Jumlah kolom dalam df_2018: 302
Jumlah kolom dalam df_2019: 302
Jumlah kolom dalam df_2020: 302
Jumlah kolom dalam df_2021: 292
Jumlah kolom dalam df_2022: 302


# **Anomali**

1. Terdapat **9 kolom hilang** di data survey **2021**
2. Bukan data survey
3. Bisa di gabung aja ke data frame **IKU**

In [6]:
df_2021.shape

(3005, 292)

In [7]:
df_2021.head()

NIM                      Nama Jenis Kelamin  Angkatan    IP   
0  11414058          Ira Indira Putri        Wanita      2014  3.61  \
1  10114063                   Ramdani          Pria      2014  3.63   
2  11414018    Kresentia Listy Andani        Wanita      2014  3.46   
3  12514021      Muhammad Iqbal Hilmi          Pria      2014  3.54   
4  10214061  Danny Eka Putra Kinarwan          Pria      2014  3.45   

                                Fakultas/Sekolah       Program Studi   
0              Sekolah Ilmu dan Teknologi Hayati  Rekayasa Pertanian  \
1  Fakultas Matematika dan Ilmu Pengetahuan Alam          Matematika   
2              Sekolah Ilmu dan Teknologi Hayati  Rekayasa Pertanian   
3   Fakultas Teknik Pertambangan dan Perminyakan    Teknik Metalurgi   
4  Fakultas Matematika dan Ilmu Pengetahuan Alam              Fisika   

  Bulan dan Tahun Wisuda          Kota   
0    2018-10-01 00:00:00      Sumedang  \
1    2019-04-01 00:00:00       Jakarta   
2    2018-10-01 00:00:00           NaN   
3    2018-07-01 00:00:00  Bandung City   
4    2018-10-01 00:00:00     Tangerang   

  Provinsi (untuk luar Indonesia cukup tuliskan nama negara)  ...   
0                                         Jawa Barat          ...  \
1                                        DKI Jakarta          ...   
2                                         Jawa Barat          ...   
3                                         Jawa Barat          ...   
4                                             Banten          ...   

  Sebutkan nama perguruan tinggi tempat Anda melanjutkan studi?   
0                                                NaN             \
1                                                NaN              
2                                                NaN              
3                                                NaN              
4                                                NaN              

  Bulan dan tahun Anda melanjutkan studi Apa alasan Anda melanjutkan studi?   
0                                    NaT                                NaN  \
1                                    NaT                                NaN   
2                                    NaT                                NaN   
3                                    NaT                                NaN   
4                                    NaT                                NaN   

  Apa bidang studi yang Anda ambil? Darimanakah sumber biaya studi Anda?   
0                               NaN                                  NaN  \
1                               NaN                                  NaN   
2                               NaN                                  NaN   
3                               NaN                                  NaN   
4                               NaN                                  NaN   

  Apakah Anda pernah bekerja sebelumnya?   
0                                    NaN  \
1                                    NaN   
2                                    NaN   
3                                    NaN   
4                                    NaN   

  Berapa lama Anda bekerja setelah lulus (pekerjaan pertama) sebelum memutuskan untuk berhenti atau tidak bekerja lagi? (dalam bulan)   
0                                                NaN                                                                                   \
1                                                NaN                                                                                    
2                                                NaN                                                                                    
3                                                NaN                                                                                    
4                                                NaN                                                                                    

  Apakah pekerjaan Anda saat itu sesuai dengan bidang kul

In [9]:
ColumnName = pd.read_excel('ColumnName.xlsx')
ColumnName['DevColumn2022']

0                    nim
1                   nama
2                    sex
3               angkatan
4                    ipk
             ...        
296         grad_support
297    workingexperience
298     delta_unemployed
299    unemployed_linear
300    unemployed_reason
Name: DevColumn2022, Length: 301, dtype: object

In [10]:
# duplicated kolom 2021
df2021 = df_2021.copy()

new_column_name = ColumnName['DevColumn2022']

for i, new_column in enumerate(new_column_name):
    df2021.rename(columns={df2021.columns[i] : new_column}, inplace=True)

IndexError: index 292 is out of bounds for axis 0 with size 292

In [11]:
# check kolom
aspek = []
bidikmisi = []
organisasi = []
kursus = []
kompetensi = []
kontribusi = []
peran = []
start = []
jobseek = []
waktu = []
pekerjaan = []
importance = []
sub_bekerja = []
sub_wirausaha = []
kesesuaian = []
jenis_w = []
grad = []
unemployed = []
lain_lain = []

for kolom in df2021.columns:
    if kolom.startswith('aspek_'):
        aspek.append(kolom)
    elif kolom.startswith('bidikmisi_'):
        bidikmisi.append(kolom)
    elif kolom.startswith('organisasi_'):
        organisasi.append(kolom)
    elif kolom.startswith('kursus_'):
        kursus.append(kolom)
    elif kolom.startswith('kompetensi_'):
        kompetensi.append(kolom)
    elif kolom.startswith('kontribusi_'):
        kontribusi.append(kolom)
    elif kolom.startswith('peran_'):
        peran.append(kolom)
    elif kolom.startswith('start_'):
        start.append(kolom)
    elif kolom.startswith('jobseek_'):
        jobseek.append(kolom)
    elif kolom.startswith('waktu'):
        waktu.append(kolom)
    elif kolom.startswith('pekerjaan_'):
        pekerjaan.append(kolom)
    elif kolom.startswith('importance_'):
        importance.append(kolom)
    elif kolom.startswith('sub_bekerja'):
        sub_bekerja.append(kolom)
    elif kolom.startswith('sub_wirausaha'):
        sub_wirausaha.append(kolom)
    elif kolom.startswith('kesesuaian'):
        kesesuaian.append(kolom)
    elif kolom.startswith('sub_wirausaha'):
        sub_wirausaha.append(kolom)
    elif kolom.startswith('jenis_w'):
        jenis_w.append(kolom)
    elif kolom.startswith('sub_wirausaha'):
        sub_wirausaha.append(kolom)
    elif kolom.startswith('grad_'):
        grad.append(kolom)
    elif kolom.startswith('unemployed_'):
        unemployed.append(kolom)
    else:
        lain_lain.append(kolom)

# hitung jumlah kolom
check_kolom = ["aspek","bidikmisi","organisasi","kursus","kompetensi","kontribusi","peran","start",
                   "jobseek","waktu","pekerjaan","importance","sub_bekerja","sub_wirausaha","kesesuaian",
                   "jenis_w","grad","unemployed","lain_lain"]

# Perulangan untuk mencetak jumlah kolom
for kolom in check_kolom:
    jumlah_kolom = len(locals()[kolom])
    print(f'Jumlah kolom {kolom} = {jumlah_kolom}')

Jumlah kolom aspek = 18
Jumlah kolom bidikmisi = 16
Jumlah kolom organisasi = 13
Jumlah kolom kursus = 19
Jumlah kolom kompetensi = 23
Jumlah kolom kontribusi = 23
Jumlah kolom peran = 23
Jumlah kolom start = 3
Jumlah kolom jobseek = 25
Jumlah kolom waktu = 6
Jumlah kolom pekerjaan = 4
Jumlah kolom importance = 14
Jumlah kolom sub_bekerja = 21
Jumlah kolom sub_wirausaha = 21
Jumlah kolom kesesuaian = 2
Jumlah kolom jenis_w = 7
Jumlah kolom grad = 0
Jumlah kolom unemployed = 0
Jumlah kolom lain_lain = 54


## **Insight:**
Kolom pada tahun **2021** yang hilang terdapat **9** kolom diantaranya,

![Alt text](image-2.png)

Nah, berarti data survey pada **Cluster** ***IKU*** tidak ada yang hilang data nya 

# **Gabung Data**

In [12]:
df_reindex = pd.concat([df_2018, df_2019, df_2020, df_2021, df_2022], ignore_index=True)
df_reindex.head()

NIM                            Nama Jenis Kelamin  Angkatan    IP   
0  NaN      Adam Nurilman Dwikadartomo          Pria    2011.0  3.30  \
1  NaN                 Adelita Febiola        Wanita    2011.0  3.50   
2  NaN  Adhietya Orlandho Putra Sunarm          Pria    2011.0  3.29   
3  NaN                    Adlan Praido          Pria    2011.0  3.39   
4  NaN               Adrianti Varadina        Wanita    2011.0  3.14   

                                    Fakultas/Sekolah Program Studi   
0  Sekolah Arsitektur, Perencanaan dan Pengembang...    Arsitektur  \
1  Sekolah Arsitektur, Perencanaan dan Pengembang...    Arsitektur   
2  Sekolah Arsitektur, Perencanaan dan Pengembang...    Arsitektur   
3  Sekolah Arsitektur, Perencanaan dan Pengembang...    Arsitektur   
4  Sekolah Arsitektur, Perencanaan dan Pengembang...    Arsitektur   

  Bulan dan Tahun Wisuda               Kota   
0                    NaN              Depok  \
1                    NaN     Bekasi Selatan   
2                    NaN            Bandung   
3                    NaN  Tangerang Selatan   
4                    NaN            Bandung   

  Provinsi (untuk luar Indonesia cukup tuliskan nama negara)  ...   
0                                         Jawa Barat          ...  \
1                                         Jawa Barat          ...   
2                                         Jawa Barat          ...   
3                                             Banten          ...   
4                                         Jawa Barat          ...   

  Sebutkan nama perguruan tinggi tempat Anda melanjutkan studi?   
0                                                NaN             \
1                                                NaN              
2                                                NaN              
3                                                NaN              
4                                                NaN              

  Bulan dan tahun Anda melanjutkan studi Apa alasan Anda melanjutkan studi?   
0                                    NaN                                NaN  \
1                                    NaN                                NaN   
2                                    NaN                                NaN   
3                                    NaN                                NaN   
4                                    NaN                                NaN   

  Apa bidang studi yang Anda ambil? Darimanakah sumber biaya studi Anda?   
0                               NaN                                  NaN  \
1                               NaN                                  NaN   
2                               NaN                                  NaN   
3                               NaN                                  NaN   
4                               NaN                                  NaN   

  Apakah Anda pernah bekerja sebelumnya?   
0                                    NaN  \
1                                    NaN   
2                                    NaN   
3                                     ya   
4                                    NaN   

  Berapa lama Anda bekerja setelah lulus (pekerjaan pertama) sebelum memutuskan untuk berhenti atau tidak bekerja lagi? (dalam bulan)   
0                                                NaN                                                                                   \
1                                                NaN                                                                                    
2                                                NaN                                                                                    
3                                               24.0                                                                                    
4                                                NaN                                                                                    

  Apakah pekerjaan Anda saat

In [13]:
Survey_IKU = df_reindex.iloc[:, [0] + [1] + [3] + list(range(88, 157)) + [-1]]

## **Note:**
dataframe final yang digunakan adalah **Survey_IKU**

# **Ganti nama kolom**

## **Note:**
Akan diubah menjadi table **DevColumn2022** sesuai data nama kolom pada data **ColumnName**

In [15]:
# ColumnName = pd.read_excel('ColumnName.xlsx')

ColumnName['DevColumn2022']

0                    nim
1                   nama
2                    sex
3               angkatan
4                    ipk
             ...        
296         grad_support
297    workingexperience
298     delta_unemployed
299    unemployed_linear
300    unemployed_reason
Name: DevColumn2022, Length: 301, dtype: object

In [16]:
# nama kolom DevColumn2022
DevColumn2022 = ColumnName['DevColumn2022'].iloc[88:157].tolist()
DevColumn2022 = ['NIM', 'Nama', 'Angkatan'] + DevColumn2022 + ['Tahun Survey']

# mengganti nama kolom baru
for i, new_name in enumerate(DevColumn2022):
    Survey_IKU.rename(columns={Survey_IKU.columns[i]: new_name}, inplace=True)

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_19860\2510377686.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Survey_IKU.rename(columns={Survey_IKU.columns[i]: new_name}, inplace=True)


In [17]:
Survey_IKU

NIM                            Nama  Angkatan   
0             NaN      Adam Nurilman Dwikadartomo    2011.0  \
1             NaN                 Adelita Febiola    2011.0   
2             NaN  Adhietya Orlandho Putra Sunarm    2011.0   
3             NaN                    Adlan Praido    2011.0   
4             NaN               Adrianti Varadina    2011.0   
...           ...                             ...       ...   
14711  13315069.0                 Reyhan Ramadhan    2015.0   
14712  10515042.0             Zulfa Hilmi Kautsar    2015.0   
14713  13615042.0           Tobias Samuel Sugandi    2015.0   
14714  10215085.0          Mohamad Reza Nurrahman    2015.0   
14715  10115026.0         Muhammad Rizki Fadillah    2015.0   

       kompetensi_masalahkompleks  kompetensi_berpikirkritis   
0                             5.0                        5.0  \
1                             4.0                        4.0   
2                             5.0                        5.0   
3                             4.0                        5.0   
4                             3.0                        4.0   
...                           ...                        ...   
14711                         4.0                        4.0   
14712                         4.0                        4.0   
14713                         4.0                        5.0   
14714                         5.0                        5.0   
14715                         4.0                        4.0   

       kompetensi_inovasi  kompetensi_managemendiri  kompetensi_bekerjatim   
0                     5.0                       4.0                    4.0  \
1                     4.0                       2.0                    2.0   
2                     4.0                       4.0                    4.0   
3                     4.0                       5.0                    2.0   
4                     4.0                       4.0                    4.0   
...                   ...                       ...                    ...   
14711                 4.0                       2.0                    2.0   
14712                 4.0                       4.0                    4.0   
14713                 2.0                       2.0                    4.0   
14714                 5.0                       5.0                    5.0   
14715                 4.0                       2.0                    2.0   

       kompetensi_bekerjaindividu  kompetensi_kecerdasanemosional  ...   
0                             3.0                             4.0  ...  \
1                             5.0                             2.0  ...   
2                             5.0                             3.0  ...   
3                             5.0                             2.0  ...   
4                             4.0                             4.0  ...   
...                           ...                             ...  ...   
14711                         4.0                             4.0  ...   
14712                         4.0                             4.0  ...   
14713                         5.0                             2.0  ...   
14714                         5.0                             5.0  ...   
14715                         4.0                             2.0  ...   

       peran_etika  peran_komunikasi  peran_penerapanbidang   
0              5.0               NaN                    4.0  \
1              4.0               NaN                    4.0   
2              5.0               NaN                    4.0   
3              3.0               NaN                    4.0   
4              4.0               NaN                    3.0   
...            ...               ...                    ...   
14711          4.0               2.0                    5.0   
14712          4.0               4.0                    4.0   
14713          4.0               5.0                    4.0   
14714          5.0              

# **Check Missing Values**

In [18]:
# Check Kolom Missing Values
def missing_values(df):

    # Mengatur opsi untuk menampilkan seluruh isi kolom
    pd.set_option('display.max_colwidth', 1)

    # Membuat list kosong untuk menyimpan informasi unik tiap kolom
    listItem = []

    # Mengiterasi tiap kolom dalam DataFrame
    for col in df.columns:
        # Memeriksa apakah ada nilai null dalam kolom
        if df[col].isnull().any():
            # Menghitung persentase missing values dengan penanganan khusus jika jumlah count non-null adalah nol
            if df[col].count() == 0:
                missing_percentage = np.nan
            else:
                missing_percentage = round((df[col].isnull().sum() / len(df)) * 100, 2)
        
            # Menambahkan informasi kolom, tipe data, jumlah unik, nilai unik, dan persentase missing values ke dalam list
            listItem.append([col, df[col].dtypes, df[col].nunique(), df[col].isnull().sum(), df[col].duplicated().sum() ,missing_percentage])

    # Membuat DataFrame dari list yang telah dibuat
    tabelMissingValues = pd.DataFrame(columns=['Column Name', 'Data Types', 'Unique Sample', 'Missing Values', 'Duplicated Values','Missing Percentage'],
                          data=listItem)

    # Menampilkan DataFrame
    return tabelMissingValues

In [19]:
missing_values(Survey_IKU)

Column Name Data Types  Unique Sample  Missing Values   
0   NIM                         float64    11873          2841            \
1   Nama                        object     14561          17               
2   Angkatan                    float64    5              25               
3   kompetensi_masalahkompleks  float64    5              2                
4   kompetensi_berpikirkritis   float64    5              2                
..                        ...       ...   ..             ..                
67  peran_analisisdata          float64    5              2                
68  peran_administrasi          float64    5              2                
69  peran_teknologi             float64    5              2                
70  peran_merancang             float64    5              2                
71  peran_bahasa_asing          float64    5              2                

    Duplicated Values  Missing Percentage  
0   2842               19.31               
1   154                0.12                
2   14710              0.17                
3   14710              0.01                
4   14710              0.01                
..    ...               ...                
67  14710              0.01                
68  14710              0.01                
69  14710              0.01                
70  14710              0.01                
71  14710              0.01                

[72 rows x 6 columns]

## **missing values di kolom NIM**

In [20]:
Survey_IKU[Survey_IKU['NIM'].isna()]

NIM                            Nama  Angkatan   
0    NaN   Adam Nurilman Dwikadartomo      2011.0    \
1    NaN   Adelita Febiola                 2011.0     
2    NaN   Adhietya Orlandho Putra Sunarm  2011.0     
3    NaN   Adlan Praido                    2011.0     
4    NaN   Adrianti Varadina               2011.0     
...   ..                 ...                  ...     
8753 NaN   NaN                            NaN         
8754 NaN   NaN                            NaN         
8755 NaN   NaN                            NaN         
8756 NaN   NaN                            NaN         
8757 NaN   NaN                            NaN         

      kompetensi_masalahkompleks  kompetensi_berpikirkritis   
0     5.0                         5.0                        \
1     4.0                         4.0                         
2     5.0                         5.0                         
3     4.0                         5.0                         
4     3.0                         4.0                         
...   ...                         ...                         
8753  4.0                         4.0                         
8754  4.0                         3.0                         
8755  3.0                         3.0                         
8756  4.0                         4.0                         
8757  4.0                         4.0                         

      kompetensi_inovasi  kompetensi_managemendiri  kompetensi_bekerjatim   
0     5.0                 4.0                       4.0                    \
1     4.0                 2.0                       2.0                     
2     4.0                 4.0                       4.0                     
3     4.0                 5.0                       2.0                     
4     4.0                 4.0                       4.0                     
...   ...                 ...                       ...                     
8753  3.0                 2.0                       4.0                     
8754  5.0                 4.0                       4.0                     
8755  4.0                 2.0                       3.0                     
8756  4.0                 5.0                       4.0                     
8757  4.0                 4.0                       3.0                     

      kompetensi_bekerjaindividu  kompetensi_kecerdasanemosional  ...   
0     3.0                         4.0                             ...  \
1     5.0                         2.0                             ...   
2     5.0                         3.0                             ...   
3     5.0                         2.0                             ...   
4     4.0                         4.0                             ...   
...   ...                         ...                             ...   
8753  5.0                         5.0                             ...   
8754  4.0                         4.0                             ...   
8755  5.0                         3.0                             ...   
8756  5.0                         5.0                             ...   
8757  4.0                         3.0                             ...   

      peran_etika  peran_komunikasi  peran_penerapanbidang   
0     5.0         NaN                4.0                    \
1     4.0         NaN                4.0                     
2     5.0         NaN                4.0                     
3     3.0         NaN                4.0                     
4     4.0         NaN                3.0                     
...   ...          ..                ...                     
8753  4.0          5.0               5.0                     
8754  3.0          4.0               3.0                     
8755  5.0          5.0               5.0                     
8756  4.0          4.0               4.0                     
8757  3.0          4.0               3.0                     

      peran_penerapanluarbidang  peran_an

In [21]:
Survey_IKU[Survey_IKU['NIM'].isna()]['Tahun Survey'].unique()

array([2018, 2020], dtype=int64)

In [22]:
Survey_IKU[Survey_IKU['NIM'].isna()]['Angkatan'].unique()

array([2011.,   nan])

In [23]:
Survey_IKU[Survey_IKU['NIM'].isna()][~((Survey_IKU[Survey_IKU['NIM'].isna()]['Tahun Survey'] == 2018))][['NIM', 'Nama', 'Angkatan']]

NIM Nama  Angkatan
8741 NaN   NaN NaN      
8742 NaN   NaN NaN      
8743 NaN   NaN NaN      
8744 NaN   NaN NaN      
8745 NaN   NaN NaN      
8746 NaN   NaN NaN      
8747 NaN   NaN NaN      
8748 NaN   NaN NaN      
8749 NaN   NaN NaN      
8750 NaN   NaN NaN      
8751 NaN   NaN NaN      
8752 NaN   NaN NaN      
8753 NaN   NaN NaN      
8754 NaN   NaN NaN      
8755 NaN   NaN NaN      
8756 NaN   NaN NaN      
8757 NaN   NaN NaN

In [24]:
df_2018.shape

(2824, 302)

### **Insight:**
1. Terdapat missing values pada data **Tahun Survey** : [2018, 2020]
2. Dengan salah satunya mahasiswa **Angkatan** : [2011] ; **Tahun Survey** : [2018]
3. Total data 2841, dimana:
    * **2824 (data NIM angkatan 2011, tahun survey 2018)** 
    * **17 (belum diketahui)**

## **missing values kolom  Nama**

In [25]:
Survey_IKU[Survey_IKU['Nama'].isna()]

NIM Nama  Angkatan  kompetensi_masalahkompleks   
8741 NaN   NaN NaN        5.0                         \
8742 NaN   NaN NaN        4.0                          
8743 NaN   NaN NaN        5.0                          
8744 NaN   NaN NaN        4.0                          
8745 NaN   NaN NaN        3.0                          
8746 NaN   NaN NaN        4.0                          
8747 NaN   NaN NaN        5.0                          
8748 NaN   NaN NaN        5.0                          
8749 NaN   NaN NaN        3.0                          
8750 NaN   NaN NaN        4.0                          
8751 NaN   NaN NaN        4.0                          
8752 NaN   NaN NaN        4.0                          
8753 NaN   NaN NaN        4.0                          
8754 NaN   NaN NaN        4.0                          
8755 NaN   NaN NaN        3.0                          
8756 NaN   NaN NaN        4.0                          
8757 NaN   NaN NaN        4.0                          

      kompetensi_berpikirkritis  kompetensi_inovasi  kompetensi_managemendiri   
8741  4.0                        4.0                 4.0                       \
8742  3.0                        3.0                 2.0                        
8743  4.0                        4.0                 3.0                        
8744  4.0                        5.0                 4.0                        
8745  3.0                        3.0                 1.0                        
8746  4.0                        4.0                 4.0                        
8747  5.0                        4.0                 3.0                        
8748  5.0                        5.0                 5.0                        
8749  1.0                        2.0                 2.0                        
8750  4.0                        3.0                 3.0                        
8751  4.0                        4.0                 3.0                        
8752  4.0                        4.0                 5.0                        
8753  4.0                        3.0                 2.0                        
8754  3.0                        5.0                 4.0                        
8755  3.0                        4.0                 2.0                        
8756  4.0                        4.0                 5.0                        
8757  4.0                        4.0                 4.0                        

      kompetensi_bekerjatim  kompetensi_bekerjaindividu   
8741  4.0                    4.0                         \
8742  3.0                    3.0                          
8743  4.0                    5.0                          
8744  4.0                    4.0                          
8745  3.0                    4.0                          
8746  4.0                    5.0                          
8747  3.0                    4.0                          
8748  5.0                    5.0                          
8749  1.0                    5.0                          
8750  3.0                    4.0                          
8751  4.0                    4.0                          
8752  3.0                    5.0                          
8753  4.0                    5.0                          
8754  4.0                    4.0                          
8755  3.0                    5.0                          
8756  4.0                    5.0                          
8757  3.0                    4.0                          

      kompetensi_kecerdasanemosional  ...  peran_etika  peran_komunikasi   
8741  4.0                             ...  5.0          5.0               \
8742  4.0                             ...  3.0          3.0                
8743  4.0                             ...  4.0          4.0                
8744  5.0                             ...  5.0          3.0                
8745  2.0                             ...  4.0          1.0                
8746  4.0          

In [26]:
Survey_IKU[Survey_IKU['Nama'].isna()][['NIM', 'Nama', 'Angkatan', 'Tahun Survey']]

NIM Nama  Angkatan  Tahun Survey
8741 NaN   NaN NaN        2020        
8742 NaN   NaN NaN        2020        
8743 NaN   NaN NaN        2020        
8744 NaN   NaN NaN        2020        
8745 NaN   NaN NaN        2020        
8746 NaN   NaN NaN        2020        
8747 NaN   NaN NaN        2020        
8748 NaN   NaN NaN        2020        
8749 NaN   NaN NaN        2020        
8750 NaN   NaN NaN        2020        
8751 NaN   NaN NaN        2020        
8752 NaN   NaN NaN        2020        
8753 NaN   NaN NaN        2020        
8754 NaN   NaN NaN        2020        
8755 NaN   NaN NaN        2020        
8756 NaN   NaN NaN        2020        
8757 NaN   NaN NaN        2020

In [27]:
Survey_IKU[Survey_IKU['Nama'].isna()]['Tahun Survey'].unique()

array([2020], dtype=int64)

In [28]:
Survey_IKU[Survey_IKU['Nama'].isna()].duplicated().sum()

0

### **Insight:**
1. dari data jumlah missing values yang belum diketahui jelas sebanyak **17** data,
2. Ternyata **17** berasal dari data **Tahun Survey** [2020] 
3. **Namun** ke-17 data ini, juga memiliki missing values pada kolom **Nama**, **NIM** dan **Angkatan**
    * data missing values ini dipastikan bukan data **duplicated**

## **missing values kolom angkatan**

In [29]:
Survey_IKU[Survey_IKU['Angkatan'].isna()]

NIM                        Nama  Angkatan   
4800  13212138.0  Meynard Danam Purwa Atmaja NaN        \
5385  10712039.0  Martin                     NaN         
5386  13212085.0  Juliver Napitupulu         NaN         
5387  13312054.0  Rizkiawan Fauzan           NaN         
5388  13212041.0  Febby Purnama Madrin       NaN         
5389  13212118.0  Fadhil Mochammad           NaN         
5390  15512080.0  Niku Guinea                NaN         
5391  13012023.0  Ricky                      NaN         
8741 NaN          NaN                        NaN         
8742 NaN          NaN                        NaN         
8743 NaN          NaN                        NaN         
8744 NaN          NaN                        NaN         
8745 NaN          NaN                        NaN         
8746 NaN          NaN                        NaN         
8747 NaN          NaN                        NaN         
8748 NaN          NaN                        NaN         
8749 NaN          NaN                        NaN         
8750 NaN          NaN                        NaN         
8751 NaN          NaN                        NaN         
8752 NaN          NaN                        NaN         
8753 NaN          NaN                        NaN         
8754 NaN          NaN                        NaN         
8755 NaN          NaN                        NaN         
8756 NaN          NaN                        NaN         
8757 NaN          NaN                        NaN         

      kompetensi_masalahkompleks  kompetensi_berpikirkritis   
4800  4.0                         4.0                        \
5385  4.0                         3.0                         
5386  4.0                         5.0                         
5387  5.0                         4.0                         
5388  4.0                         4.0                         
5389  4.0                         4.0                         
5390  4.0                         4.0                         
5391  5.0                         5.0                         
8741  5.0                         4.0                         
8742  4.0                         3.0                         
8743  5.0                         4.0                         
8744  4.0                         4.0                         
8745  3.0                         3.0                         
8746  4.0                         4.0                         
8747  5.0                         5.0                         
8748  5.0                         5.0                         
8749  3.0                         1.0                         
8750  4.0                         4.0                         
8751  4.0                         4.0                         
8752  4.0                         4.0                         
8753  4.0                         4.0                         
8754  4.0                         3.0                         
8755  3.0                         3.0                         
8756  4.0                         4.0                         
8757  4.0                         4.0                         

      kompetensi_inovasi  kompetensi_managemendiri  kompetensi_bekerjatim   
4800  5.0                 3.0                       3.0                    \
5385  3.0                 3.0                       3.0                     
5386  5.0                 4.0                       5.0                     
5387  5.0                 4.0                       5.0                     
5388  5.0                 4.0                       5.0                     
5389  4.0                 3.0                       4.0                     
5390  4.0                 4.0                       4.0                     
5391  5.0                 5.0                       3.0                     
8741  4.0                 4.0                       4.0                     
8742  3.0                 2.0                       3.0                     
8743  4.0         

In [30]:
Survey_IKU[Survey_IKU['Angkatan'].isna()][~((Survey_IKU[Survey_IKU['Angkatan'].isna()]['Tahun Survey'] == 2020))]

NIM                        Nama  Angkatan   
4800  13212138.0  Meynard Danam Purwa Atmaja NaN        \
5385  10712039.0  Martin                     NaN         
5386  13212085.0  Juliver Napitupulu         NaN         
5387  13312054.0  Rizkiawan Fauzan           NaN         
5388  13212041.0  Febby Purnama Madrin       NaN         
5389  13212118.0  Fadhil Mochammad           NaN         
5390  15512080.0  Niku Guinea                NaN         
5391  13012023.0  Ricky                      NaN         

      kompetensi_masalahkompleks  kompetensi_berpikirkritis   
4800  4.0                         4.0                        \
5385  4.0                         3.0                         
5386  4.0                         5.0                         
5387  5.0                         4.0                         
5388  4.0                         4.0                         
5389  4.0                         4.0                         
5390  4.0                         4.0                         
5391  5.0                         5.0                         

      kompetensi_inovasi  kompetensi_managemendiri  kompetensi_bekerjatim   
4800  5.0                 3.0                       3.0                    \
5385  3.0                 3.0                       3.0                     
5386  5.0                 4.0                       5.0                     
5387  5.0                 4.0                       5.0                     
5388  5.0                 4.0                       5.0                     
5389  4.0                 3.0                       4.0                     
5390  4.0                 4.0                       4.0                     
5391  5.0                 5.0                       3.0                     

      kompetensi_bekerjaindividu  kompetensi_kecerdasanemosional  ...   
4800  4.0                         4.0                             ...  \
5385  3.0                         3.0                             ...   
5386  5.0                         4.0                             ...   
5387  4.0                         2.0                             ...   
5388  5.0                         4.0                             ...   
5389  4.0                         3.0                             ...   
5390  4.0                         4.0                             ...   
5391  5.0                         5.0                             ...   

      peran_etika  peran_komunikasi  peran_penerapanbidang   
4800  5.0          4.0               3.0                    \
5385  4.0          4.0               4.0                     
5386  5.0          2.0               2.0                     
5387  5.0          2.0               5.0                     
5388  5.0          4.0               4.0                     
5389  4.0          4.0               4.0                     
5390  4.0          4.0               4.0                     
5391  5.0          5.0               5.0                     

      peran_penerapanluarbidang  peran_analisisdata  peran_administrasi   
4800  3.0                        4.0                 4.0                 \
5385  4.0                        5.0                 4.0                  
5386  2.0                        2.0                 2.0                  
5387  4.0                        4.0                 5.0                  
5388  4.0                        4.0                 4.0                  
5389  4.0                        4.0                 4.0                  
5390  4.0                        4.0                 4.0                  
5391  5.0                        5.0                 5.0                  

      peran_teknologi  peran_merancang  peran_bahasa_asing  Tahun Survey  
4800  4.0              2.0              5.0                 2019          
5385  4.0              4.0              5.0                 2019          
5386  3.0              2.0              2.0                 2019          
5387  5.0              5.0         

In [31]:
Survey_IKU[Survey_IKU['Angkatan'].isna()][~((Survey_IKU[Survey_IKU['Angkatan'].isna()]['Tahun Survey'] == 2020))][['NIM', 'Nama', 'Angkatan', 'Tahun Survey']]

NIM                        Nama  Angkatan  Tahun Survey
4800  13212138.0  Meynard Danam Purwa Atmaja NaN        2019        
5385  10712039.0  Martin                     NaN        2019        
5386  13212085.0  Juliver Napitupulu         NaN        2019        
5387  13312054.0  Rizkiawan Fauzan           NaN        2019        
5388  13212041.0  Febby Purnama Madrin       NaN        2019        
5389  13212118.0  Fadhil Mochammad           NaN        2019        
5390  15512080.0  Niku Guinea                NaN        2019        
5391  13012023.0  Ricky                      NaN        2019

In [32]:
25 - 8

17

### **Insight:**
1. Total terdapat **25** data missing values
2. **17** data diantaranya, tidak memiliki **NIM, Nama, Angkatan**, namun diketahui data **Tahun Survey 2020**
2. Sisa **8** data dari memiliki kolom **NIM**, **Nama**, **Tahun Survey** [2019]

In [33]:
missing_values(Survey_IKU)

Column Name Data Types  Unique Sample  Missing Values   
0   NIM                         float64    11873          2841            \
1   Nama                        object     14561          17               
2   Angkatan                    float64    5              25               
3   kompetensi_masalahkompleks  float64    5              2                
4   kompetensi_berpikirkritis   float64    5              2                
..                        ...       ...   ..             ..                
67  peran_analisisdata          float64    5              2                
68  peran_administrasi          float64    5              2                
69  peran_teknologi             float64    5              2                
70  peran_merancang             float64    5              2                
71  peran_bahasa_asing          float64    5              2                

    Duplicated Values  Missing Percentage  
0   2842               19.31               
1   154                0.12                
2   14710              0.17                
3   14710              0.01                
4   14710              0.01                
..    ...               ...                
67  14710              0.01                
68  14710              0.01                
69  14710              0.01                
70  14710              0.01                
71  14710              0.01                

[72 rows x 6 columns]

## **missing values kolom IKU**

In [34]:
for index, row in missing_values(Survey_IKU)[['Column Name', 'Missing Percentage']].iterrows():
    x = row['Column Name']
    y = row['Missing Percentage']
    
    if y > 3:
        print(x, y)

NIM 19.31
kompetensi_komunikasi 19.2
kontribusi_komunikasi 19.2
peran_komunikasi 19.2


In [35]:
missing_values(Survey_IKU)[(missing_values(Survey_IKU)['Column Name'] == 'kompetensi_komunikasi') | (missing_values(Survey_IKU)['Column Name'] == 'kontribusi_komunikasi')]

Column Name Data Types  Unique Sample  Missing Values   
18  kompetensi_komunikasi  float64    5              2826            \
41  kontribusi_komunikasi  float64    5              2826             

    Duplicated Values  Missing Percentage  
18  14710              19.2                
41  14710              19.2

### **Insight:**
1. Terdapat dua kolom **IKU** melebihi batas missing values yaitu kolom **kompetensi_komunikasi** dan **kontribusi_komunikasi**

In [36]:
Survey_IKU[Survey_IKU['kompetensi_komunikasi'].isnull()][['NIM', 'Nama', 'Angkatan', 'Tahun Survey']]

NIM                            Nama  Angkatan  Tahun Survey
0    NaN          Adam Nurilman Dwikadartomo      2011.0    2018        
1    NaN          Adelita Febiola                 2011.0    2018        
2    NaN          Adhietya Orlandho Putra Sunarm  2011.0    2018        
3    NaN          Adlan Praido                    2011.0    2018        
4    NaN          Adrianti Varadina               2011.0    2018        
...   ..                        ...                  ...     ...        
2821 NaN          Yosafat Marthin Samosir         2011.0    2018        
2822 NaN          Yuka Himawan                    2011.0    2018        
2823 NaN          Yusuf Fajar Ardiana             2011.0    2018        
4052  13412082.0  Heni Pratiwi                    2012.0    2019        
8529  15013004.0  Ahmad Sulaiman                  2013.0    2020        

[2826 rows x 4 columns]

In [37]:
Survey_IKU[Survey_IKU['kontribusi_komunikasi'].isnull()][['NIM', 'Nama', 'Angkatan', 'Tahun Survey']]

NIM                            Nama  Angkatan  Tahun Survey
0    NaN          Adam Nurilman Dwikadartomo      2011.0    2018        
1    NaN          Adelita Febiola                 2011.0    2018        
2    NaN          Adhietya Orlandho Putra Sunarm  2011.0    2018        
3    NaN          Adlan Praido                    2011.0    2018        
4    NaN          Adrianti Varadina               2011.0    2018        
...   ..                        ...                  ...     ...        
2821 NaN          Yosafat Marthin Samosir         2011.0    2018        
2822 NaN          Yuka Himawan                    2011.0    2018        
2823 NaN          Yusuf Fajar Ardiana             2011.0    2018        
4052  13412082.0  Heni Pratiwi                    2012.0    2019        
8529  15013004.0  Ahmad Sulaiman                  2013.0    2020        

[2826 rows x 4 columns]

# **Ubah Nama kolom --> DevColumn2022**

In [39]:
ColumnName['DevColumn2022'].iloc[88:157]

88     kompetensi_masalahkompleks
89     kompetensi_berpikirkritis 
90     kompetensi_inovasi        
91     kompetensi_managemendiri  
92     kompetensi_bekerjatim     
               ...               
152    peran_analisisdata        
153    peran_administrasi        
154    peran_teknologi           
155    peran_merancang           
156    peran_bahasa_asing        
Name: DevColumn2022, Length: 69, dtype: object

In [40]:
# nama kolom DevColumn2022
DevColumn2022 = ColumnName['DevColumn2022'].iloc[88:157].tolist()
DevColumn2022 = ['NIM', 'Nama', 'Angkatan'] + DevColumn2022 + ['Tahun Survey']

# mengganti nama kolom baru
for i, new_name in enumerate(DevColumn2022):
    Survey_IKU.rename(columns={Survey_IKU.columns[i]: new_name}, inplace=True)

C:\Users\LENOVO\AppData\Local\Temp\ipykernel_19860\2510377686.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Survey_IKU.rename(columns={Survey_IKU.columns[i]: new_name}, inplace=True)


In [42]:
Survey_IKU.iloc[:, [1] + list(range(3, 73))]

Nama  kompetensi_masalahkompleks   
0      Adam Nurilman Dwikadartomo      5.0                         \
1      Adelita Febiola                 4.0                          
2      Adhietya Orlandho Putra Sunarm  5.0                          
3      Adlan Praido                    4.0                          
4      Adrianti Varadina               3.0                          
...                  ...               ...                          
14711  Reyhan Ramadhan                 4.0                          
14712  Zulfa Hilmi Kautsar             4.0                          
14713  Tobias Samuel Sugandi           4.0                          
14714  Mohamad Reza Nurrahman          5.0                          
14715  Muhammad Rizki Fadillah         4.0                          

       kompetensi_berpikirkritis  kompetensi_inovasi   
0      5.0                        5.0                 \
1      4.0                        4.0                  
2      5.0                        4.0                  
3      5.0                        4.0                  
4      4.0                        4.0                  
...    ...                        ...                  
14711  4.0                        4.0                  
14712  4.0                        4.0                  
14713  5.0                        2.0                  
14714  5.0                        5.0                  
14715  4.0                        4.0                  

       kompetensi_managemendiri  kompetensi_bekerjatim   
0      4.0                       4.0                    \
1      2.0                       2.0                     
2      4.0                       4.0                     
3      5.0                       2.0                     
4      4.0                       4.0                     
...    ...                       ...                     
14711  2.0                       2.0                     
14712  4.0                       4.0                     
14713  2.0                       4.0                     
14714  5.0                       5.0                     
14715  2.0                       2.0                     

       kompetensi_bekerjaindividu  kompetensi_kecerdasanemosional   
0      3.0                         4.0                             \
1      5.0                         2.0                              
2      5.0                         3.0                              
3      5.0                         2.0                              
4      4.0                         4.0                              
...    ...                         ...                              
14711  4.0                         4.0                              
14712  4.0                         4.0                              
14713  5.0                         2.0                              
14714  5.0                         5.0                              
14715  4.0                         2.0                              

       kompetensi_pengambilankeputusan  kompetensi_negosiasi  ...   
0      4.0                              4.0                   ...  \
1      4.0                              2.0                   ...   
2      5.0                              3.0                   ...   
3      4.0                              3.0                   ...   
4      5.0                              3.0                   ...   
...    ...                              ...                   ...   
14711  4.0                              1.0                   ...   
14712  4.0                              2.0                   ...   
14713  2.0                              1.0                   ...   
14714  5.0                              5.0                   ...   
14715  2.0                              2.0                   ...   

       peran_etika  peran_komunikasi  peran_penerapanbidang   
0      5.0         NaN                4.0                    \
1      4.0         NaN                4.0         

In [43]:
Survey_IKU.iloc[:, [1] + list(range(3, 73))].pivot_table(
    index='Tahun Survey',
    columns='Nama'
)

kompetensi_adaptasi                                               
Nama          ' Aqil ' Athoillah 'Isya' Nur Fatnaya A A Dharma Adi Pratidnya   
Tahun Survey                                                                   
2018         NaN                 NaN                NaN                       \
2019         NaN                 NaN                NaN                        
2020          3.0                 3.0                4.0                       
2021         NaN                 NaN                NaN                        
2022         NaN                 NaN                NaN                        

                                                                  
Nama         A Bahrul Ulum Azmi A Fine Hastuti A Kayan Manggala   
Tahun Survey                                                      
2018          4.0               NaN             5.0              \
2019         NaN                NaN            NaN                
2020         NaN                NaN            NaN                
2021         NaN                 4.0           NaN                
2022         NaN                NaN            NaN                

                                                                          
Nama         A Syauqi Mirza Fakhri A Young Kim A. Farid Khoirudin Muhyi   
Tahun Survey                                                              
2018         NaN                   NaN         NaN                       \
2019         NaN                   NaN         NaN                        
2020         NaN                   NaN         NaN                        
2021         NaN                    5.0         5.0                       
2022          5.0                  NaN         NaN                        

                                  ...  peran_teknologi                       
Nama         A. Pradivta Nugraha  ... Zulhaerana Bahar Zulhendri Eka Putra   
Tahun Survey                      ...                                        
2018         NaN                  ... NaN              NaN                  \
2019          3.0                 ...  3.0             NaN                   
2020         NaN                  ... NaN               5.0                  
2021         NaN                  ... NaN              NaN                   
2022         NaN                  ... NaN              NaN                   

                                                           
Nama         Zulia Manalu Zulis Septian Zulkifli Yuanata   
Tahun Survey                                               
2018         NaN          NaN           NaN               \
2019          4.0          3.0          NaN                
2020         NaN          NaN            4.0               
2021         NaN          NaN           NaN                
2022         NaN          NaN           NaN                

                                                                             
Nama         Zulnio Rizqy Hafizh Banjaransari Zulva Fachrina Zumrotun Nisa   
Tahun Survey                                                                 
2018         NaN                              NaN            NaN            \
2019         NaN                              NaN            NaN             
2020          4.0                              5.0           NaN             
2021         NaN                              NaN             4.0            
2022         NaN                              NaN            NaN             

                                    
Nama         Zunaiza astrid amalia  
Tahun Survey                        
2018          4.0    NaN            
2019         NaN     NaN            
2020         NaN     NaN            
2021         NaN      5.0           
2022         NaN     NaN            

[5 rows x 996231 columns]

# **Ubah nama kolom --> Cluster**

In [ ]:
# Survey_IKU_2 = Survey_IKU.copy()

In [ ]:
# nama_kolom['Cluster'].iloc[88:157]

In [ ]:
# # nama kolom baru
# Cluster = nama_kolom['Cluster'].iloc[88:157].tolist()
# Cluster = ['NIM', 'Nama', 'Angkatan'] + Cluster + ['Tahun Survey']

# # ganti nama kolom baru
# for i, new_name in enumerate(Cluster):
#     Survey_IKU_2.rename(columns={Survey_IKU.columns[i] : new_name}, inplace=True)

In [ ]:
# Survey_IKU_2

In [ ]:
# Survey_IKU_2.iloc[:, [1] + list(range(3,73))]

In [ ]:
# Survey_IKU_2.iloc[:, [1] + list(range(3,73))].pivot_table(
#     index='Tahun Survey',
#     columns= 'Nama'
# ).reset_index()

In [ ]:
# transpose nama kolom jadi baris --> buat kolom cluster
# Survey_IKU_2 = Survey_IKU_2.transpose().reset_index()

# # ubah nama kolom jadi bilangan
# new_columns_name = range(0, len(Survey_IKU_2.columns)-1)

# Survey_IKU_2.rename(columns=dict(zip(Survey_IKU_2.columns, new_columns_name)))

# Survey_IKU_2

In [ ]:
# # Check Kolom Missing Values
# def missing_values(x):

#     # Mengatur opsi untuk menampilkan seluruh isi kolom
#     pd.set_option('display.max_colwidth', 1)

#     # Membuat list kosong untuk menyimpan informasi unik tiap kolom
#     listItem = []

#     # Mengiterasi tiap kolom dalam DataFrame
#     for col in x.columns:
#         # Memeriksa apakah ada nilai null dalam kolom
#         if x[col].isnull().any():
#             # Menghitung persentase missing values dengan penanganan khusus jika jumlah count non-null adalah nol
#             if x[col].count() == 0:
#                 missing_percentage = np.nan
#             else:
#                 missing_percentage = round((x[col].isnull().sum() / len(x)) * 100, 2)
        
#             # Menambahkan informasi kolom, tipe data, jumlah unik, nilai unik, dan persentase missing values ke dalam list
#             listItem.append([col, x[col].dtypes, x[col].nunique(), x[col].unique(), x[col].isnull().sum() ,missing_percentage])

#     # Membuat DataFrame dari list yang telah dibuat
#     tabelMissingValues = pd.DataFrame(columns=['Column Name', 'Data Types', 'Number of Unique', 'Unique Sample', 'Jumlah Missing', 'Missing Percentage'],
#                           data=listItem)

#     # Menampilkan DataFrame
#     return tabelMissingValues

In [ ]:
# missing_values(Survey_IKU)

# **Duplicated Values**

### **Duplicated Values: NIM**

In [44]:
missing_values(Survey_IKU)

Column Name Data Types  Unique Sample  Missing Values   
0   NIM                         float64    11873          2841            \
1   Nama                        object     14561          17               
2   Angkatan                    float64    5              25               
3   kompetensi_masalahkompleks  float64    5              2                
4   kompetensi_berpikirkritis   float64    5              2                
..                        ...       ...   ..             ..                
67  peran_analisisdata          float64    5              2                
68  peran_administrasi          float64    5              2                
69  peran_teknologi             float64    5              2                
70  peran_merancang             float64    5              2                
71  peran_bahasa_asing          float64    5              2                

    Duplicated Values  Missing Percentage  
0   2842               19.31               
1   154                0.12                
2   14710              0.17                
3   14710              0.01                
4   14710              0.01                
..    ...               ...                
67  14710              0.01                
68  14710              0.01                
69  14710              0.01                
70  14710              0.01                
71  14710              0.01                

[72 rows x 6 columns]

In [53]:
Survey_IKU['NIM'][Survey_IKU['NIM'].notnull()].duplicated().any()

True

In [65]:
# Membuat DataFrame baru dengan NIM yang tidak null
non_null_NIM = Survey_IKU[Survey_IKU['NIM'].notnull()]

# Mengidentifikasi baris dengan NIM yang sama
duplicate_NIM_rows = non_null_NIM[non_null_NIM.duplicated(subset='NIM', keep=False)]

# Menampilkan baris-baris dengan NIM yang sama
# print(duplicate_NIM_rows)
duplicate_NIM_rows[['NIM', 'Nama', 'Angkatan', 'Tahun Survey']]

NIM              Nama  Angkatan  Tahun Survey
6041   17513501.0  Geri Primera      2013.0    2020        
9388   17513501.0  Gery Primera      2014.0    2021        
10894  12014501.0  Septiani Perdana  2014.0    2021        
11886  12014501.0  Septiani Perdana  2015.0    2022

### **Duplicated Values : Nama**

In [58]:
Survey_IKU['Nama'].duplicated().sum()

154

In [57]:
# Membuat DataFrame baru dengan Nama yang tidak null
non_null_Nama = Survey_IKU[Survey_IKU['Nama'].notnull()]

# Mengidentifikasi baris dengan Nama yang sama
duplicate_Nama_rows = non_null_Nama[non_null_Nama.duplicated(subset='Nama', keep=False)]

# Menampilkan baris-baris dengan Nama yang sama
duplicate_Nama_rows

NIM            Nama  Angkatan  kompetensi_masalahkompleks   
25    NaN          Christian       2011.0    4.0                         \
91    NaN          Sherly          2011.0    4.0                          
252   NaN          Eva Afifah      2011.0    3.0                          
433   NaN          Muhammad Ilham  2011.0    4.0                          
604   NaN          Nurul Lathifah  2011.0    4.0                          
...    ..                     ...     ...    ...                          
14383  18315006.0  Aditya Pratama  2015.0    5.0                          
14457  15715016.0  Abraham         2015.0    4.0                          
14477  13015050.0  Christian       2015.0    5.0                          
14665  11215003.0  Stanley         2015.0    4.0                          
14682  13615039.0  Daniel          2015.0    4.0                          

       kompetensi_berpikirkritis  kompetensi_inovasi   
25     5.0                        4.0                 \
91     3.0                        3.0                  
252    4.0                        4.0                  
433    4.0                        4.0                  
604    4.0                        4.0                  
...    ...                        ...                  
14383  5.0                        5.0                  
14457  4.0                        4.0                  
14477  5.0                        4.0                  
14665  4.0                        4.0                  
14682  4.0                        4.0                  

       kompetensi_managemendiri  kompetensi_bekerjatim   
25     4.0                       4.0                    \
91     4.0                       3.0                     
252    4.0                       4.0                     
433    4.0                       4.0                     
604    3.0                       2.0                     
...    ...                       ...                     
14383  4.0                       4.0                     
14457  4.0                       4.0                     
14477  5.0                       5.0                     
14665  4.0                       4.0                     
14682  4.0                       4.0                     

       kompetensi_bekerjaindividu  kompetensi_kecerdasanemosional  ...   
25     5.0                         3.0                             ...  \
91     4.0                         4.0                             ...   
252    5.0                         2.0                             ...   
433    4.0                         4.0                             ...   
604    4.0                         3.0                             ...   
...    ...                         ...                             ...   
14383  4.0                         4.0                             ...   
14457  5.0                         4.0                             ...   
14477  5.0                         4.0                             ...   
14665  4.0                         4.0                             ...   
14682  4.0                         4.0                             ...   

       peran_etika  peran_komunikasi  peran_penerapanbidang   
25     5.0         NaN                5.0                    \
91     5.0         NaN                5.0                     
252    3.0         NaN                4.0                     
433    4.0         NaN                4.0                     
604    3.0         NaN                2.0                     
...    ...          ..                ...                     
14383  5.0          5.0               5.0                     
14457  4.0          4.0               4.0                     
14477  5.0          4.0               4.0                     
14665  4.0          4.0               4.0                     
14682  5.0          4.0               5.0                     

       peran_penerapanluarbidang  peran_analisisdata  peran_administrasi   
25     5.0           

In [60]:
len(duplicate_Nama_rows['Nama'].unique())

98

In [61]:
duplicate_Nama_rows['Nama'].unique()

array(['Christian', 'Sherly', 'Eva Afifah', 'Muhammad Ilham',
       'Nurul Lathifah', 'Muhammad Luthfi', 'Steven', 'Lani Diana',
       'Rina Yuliana', 'Ferry Fernando', 'Muhammad Ihsan',
       'Muhammad Irfan', 'Taufik Hidayat', 'Putri Indah Lestari',
       'Muhammad Ilham Akbar', 'Gilang Ramadhan', 'Muhammad Farhan',
       'Michael', 'Muhammad Afif', 'Fadel Muhammad', 'Siti Aisyah',
       'Aditya Pratama', 'Naufal Muhammad', 'Stanley', 'Vinsensius',
       'Arief Rahman', 'Daniel', 'Erwin', 'Muhammad Ikhsan',
       'Fakhrul Hidayat', 'Abdul Azis', 'Steven Lie', 'Ali Akbar',
       'Ilham', 'Anthony', 'Cynthia', 'Kevin', 'Angelia', 'Nicholas',
       'Nurul Fatimah', 'Richard', 'Muhammad Iqbal', 'Annisa Amalia',
       'Muhammad Arif Hidayat', 'Slamet Riyadi', 'Fachri Muhammad',
       'Erika', 'Khairunnisa', 'Kevin Christian', 'Nurul Aini',
       'Muhammad Naufal', 'Muhammad Rizki', 'Sri Wahyuni', 'Robert',
       'Irwan', 'Ricky', 'Mohammad Fajar', 'Calvin',
       'Muhammad 

In [66]:
Survey_IKU[Survey_IKU['Nama'] == 'Steven'][['NIM', 'Nama', 'Angkatan', 'Tahun Survey']]

NIM    Nama  Angkatan  Tahun Survey
676   NaN          Steven  2011.0    2018        
2714  NaN          Steven  2011.0    2018        
7328   13213051.0  Steven  2013.0    2020        
12402  13415065.0  Steven  2015.0    2022        
14035  13015078.0  Steven  2015.0    2022

### **Insight:**
1. Ada **98** duplicated values --> ini sudah bersih dari **missing values**
2. Dari **98** diatas --> perlu di check satu-satu, ***apakah data tersebut pada **NIM** yang **sama*****, jika iya perlu tindakan? 